io-sim/src/Control/Monad/IOSim/IORef.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE BangPatterns  #-}
+{-# LANGUAGE MagicHash     #-}
+{-# LANGUAGE TypeFamilies  #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Control.Monad.IOSim.IORef where
+
+import Control.Monad.Class.MonadIORef
+import Control.Monad.Class.MonadST
+import Control.Monad.IOSim.Types
+import GHC.Exts
+import GHC.ST
+import GHC.STRef
+
+newtype IOSimRef s a = IORef (STRef s a)
+
+instance MonadIORef (IOSim s) where
+  type IORef (IOSim s) = IOSimRef s
+  newIORef v = stToIO (newSTRef v) >>= \ var -> return (IORef var)
+  readIORef (IORef v) = stToIO (readSTRef v)
+  writeIORef (IORef var) v = stToIO (writeSTRef var v)
+  modifyIORef ref f = readIORef ref >>= writeIORef ref . f
+  modifyIORef' ref f = do
+    x <- readIORef ref
+    let x' = f x
+    x' `seq` writeIORef ref x'
+  atomicModifyIORef ref f = do
+    (_old, (_new, res)) <- atomicModifyIORef2 ref f
+    pure res
+  atomicModifyIORef' = Control.Monad.IOSim.IORef.atomicModifyIORef'
+  atomicWriteIORef ref a = do
+    _ <- atomicSwapIORef ref a
+    pure ()
+
+atomicModifyIORef2Lazy :: IORef (IOSim s) a -> (a -> (a,b)) -> IOSim s (a, (a, b))
+atomicModifyIORef2Lazy (IORef (STRef r#)) f = stToIO $
+  ST (\s -> case atomicModifyMutVar2# r# f s of
+              (# s', old, res #) -> (# s', (old, res) #))
```

Comment:
Since `IORefs` are `STRef`s, we can copy the code from `base`. However, this is probably overkill -- @coot, do you have pointers for how I should proceed?

Reply:
Also, these implementations might change slightly between GHC releases. I see that GHA is failing because GHCs before `9.8` don't have access to an `atomicSwapMutVar#` function